rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,209 @@
+```
+The synthesizer will enumerate candidates derived from `NT_Bool` in the following order.
+```
+y == y
+y == LOOP_ENTRY(y)
+y == 1
+...
+1 <= y + 1
+...
+```
+The synthesizer then verifies with CBMC if the candidate is an inductive invariant that can be used to prove the post-condition `y == 0`.
+For example, the candidate `y == y` is verified to be an inductive invariant, but cannot be used to prove the post-condition `y == 0`.
+The candidate `y == 1` is not inductive.
+The synthesizer rejects all candidates until it finds the candidate `1 <= y + 1`, which can be simplified to `y >= 0`.
+`y >= 0` is an inductive invariant that can be used to prove the post-condition.
+So the synthesizer will return `y >= 0` and apply it in the goto model to get `main_synthesized.goto`.
+Then Kani will call `cbmc` on `main_synthesized.goto` to verify the program with the synthesized loop contracts. 
+
+## Rationale and alternatives
+
+- Different candidate space.
+The candidate grammar introduced above now only contains a restricted set of operators, which works well for array-manipulating programs with only pointer-checks instrumented by `goto-instrument`, but probably not enough for other user-written checks.
+We may want to include array-indexing, pointer-dereference, or other arithmetic operators in the candidate grammar for synthesizing a larger set of loop invariants.
+However, there is a trade-off between the size of candidate we enumerate and the running time of the enumeration.
+We will collect more data to decide what operators we should include in the candidate grammar.
+Once we decide more kinds of candidate grammars, we will provide users options to choose which candidate grammar they want to use.
+
+## Open questions
+
+**How does the synthesizer work with unwinding numbers?**
+There may exist some loops for which the synthesizer cannot find loop contracts, but some small unwinding numbers are enough to cover all executions of the loops.
+In this case, we may want to unwind some loops in the program while synthesizing loop contracts for other loops.
+It requires us to have a way to identify and specify which loops we want to unwind. 
+
+In C programs, we identify loops by **loop id**, which is a pair (function name, loop number).
+However, it Rust programs, loops are usually in library functions such as `Iterator::for_each`.
+And a library function may be called from different places in the program.
+We may want to unwind the loop in some calls but not in other calls.
+
+**How do we output the synthesized loop contracts?**
+To better earn users' trust, we want to be able to report what loop contracts we synthesized and used to verify the given programs.
+Now `goto-synthesizer` can dump the synthesized loop contracts into a json file.
+Here is an example of the dumped loop contracts.
+It contains the location of source files of the loops, the synthesized loop invariants and loop assigns for loops identified by loop numbers.
```

Comment:
"loops" appears 5 times in this sentences, but "loop assigns" doesn't mean anything. Can you rewrite this sentence? For example, I believe "loop assigns" can be replaced with "assign clauses".